Public/Src/Cache/ContentStore/Distributed/NuCache/ContentLocationDatabase.cs

```diff
@@ -659,43 +659,25 @@ internal bool ForceCacheFlush(OperationContext context, ContentLocationDatabaseC
                 return false;
             }
 
-            // Ensure only one thread can start the flush
-            if (Interlocked.CompareExchange(ref _isFlushInProgress, 1, 0) == 1)
+            bool renewed = false;
+            if (_flushTask.IsCompleted)
             {
-                if (blocking)
+                lock (_flushTaskLock)
                 {
-                    // If the flush is blocking, then we need to wait until it is over for correctness
-                    _flushEvent.Wait();
-                    return true;
+                    if (_flushTask.IsCompleted)
+                    {
+                        _flushTask = forceCacheFlushAsync(context, counter);
+                        renewed = true;
+                    }
                 }
-
-                return false;
             }
 
-            var task = forceCacheFlushAsync(context, counter).ContinueWith(_ =>
-            {
-                // Release all threads waiting for the flush to finish
-                _flushEvent.Set();
-
-                // If a blocking flush gets interleaved with this code, the blocking flush could potentially be forced
-                // to wait until another thread comes along and effectively performs the flush. Not much we can do
-                // about this, as the alternative is just as bad or worse.
-
-                _flushEvent.Reset();
-
-                Interlocked.Exchange(ref _isFlushInProgress, 0);
-            });
-
             if (blocking)
             {
-                task.GetAwaiter().GetResult();
-            }
-            else
-            {
-                task.FireAndForget(context);
+                _flushTask.GetAwaiter().GetResult();
             }
 
-            return true;
+            return renewed || blocking;

```

Comment:
Right now the logic is interesting.
- If memory cache is off the result is off even when blocking is true
- if blocking is true then the result is true regardless whether we flushed or not
- if blocking is false, then the result indicates whether we actually flushed or not.

This kind of complicated imo:)

Reply:
This is exactly what it's supposed to be. The result is supposed to be whether the call effectively performed a flush, modulo blocking flushes (in which case we consider it the same as we are waiting for it to finish). The return value is only used for testing

---
In reply to: [342705873](https://github.com/microsoft/BuildXL/pull/1163#discussion_r342705873) [](ancestors = 342705873)

Reply:
Given that the result is used in tests only and the false was never checked, I doubt that the current logic is good.

For instance, you can change this to return true only when it is blocked and it was flushed and the system will work as it is right now.

---
In reply to: [342730347](https://github.com/microsoft/BuildXL/pull/1163#discussion_r342730347) [](ancestors = 342730347,342705873)

Reply:
I actually think that the result should be true when the flush happened synchronously and when actually happened.

---
In reply to: [342731585](https://github.com/microsoft/BuildXL/pull/1163#discussion_r342731585) [](ancestors = 342731585,342730347,342705873)